File: frontend/src/types/strategy.ts
// 与 backend/aitrade/models/strategy.py 对齐
// Rule-based strategy types for frontend ↔ backend communication

/**
 * 可用信号源的元信息（`GET /api/strategy/sources` 列表元素）。
 *
 * 前端用 `param_spec` 动态渲染信号参数表单，不需要硬编码各信号源的字段。
 */
export interface SignalSourceInfo {
  /** 信号源唯一名称，如 "etf_momentum" */
  name: string
  /** 信号源简短说明，展示在 Select 选项中 */
  description: string
  /**
   * 可配置参数规格；null 表示无额外参数。
   * 键为参数名，值含类型（int/float/str/list[str]）、默认值与显示标签。
   */
  param_spec: Record<string, { type?: string; default?: unknown; label?: string }> | null
}

/**
 * 规则策略回测的成本假设（映射后端 StrategyCostRequest）。
 *
 * 各字段以小数表示比率（如 0.0003 = 万3）。
 */
export interface StrategyCost {
  /** 单边佣金率，如 0.0003 */
  commission_rate: number
  /** 卖出印花税率，如 0.0005 */
  stamp_duty: number
  /** 成交滑点比率，如 0.0005 */
  slippage: number
  /** 是否启用 T+1 限制（当日买入不可当日卖出） */
  t_plus1: boolean
}

/**
 * 启动单次规则策略回测的请求体（`POST /api/strategy/backtest/run`）。
 *
 * 与后端 `StrategyBacktestRequest` Pydantic 模型对齐。
 */
export interface StrategyBacktestRequest {
  /** 信号源名称，需与 `GET /api/strategy/sources` 返回的 name 一致 */
  signal_source: string
  /** 信号源参数，必须包含 universe（标的池列表） */
  signal_params: Record<string, unknown>
  /** 策略名称（可选，留空则由后端自动生成） */
  strategy_name?: string
  /** 策略参数，如 top_k / n_drop / rebalance_freq */
  strategy_params: Record<string, unknown>
  /** 行情周期，如 "d"（日线）；缺省为 "d" */
  interval?: string
  /** 回测起始日期 YYYY-MM-DD（含） */
  start: string
  /** 回测结束日期 YYYY-MM-DD（含） */
  end: string
  /** 初始资金（元）；缺省 1000000 */
  capital?: number
  /** 成本假设；缺省使用后端 A 股默认值 */
  cost?: StrategyCost
}

/**
 * 启动参数网格扫描的请求体（`POST /api/strategy/sweep/run`）。
 *
 * 在 `StrategyBacktestRequest` 基础上追加 `grid` 网格列表，每行为一个参数覆盖组合。
 *
 * @example
 * ```json
 * { "strategy_params": { "top_k": 3 } }
 * { "strategy_params": { "top_k": 5 } }
 * ```
 */
export interface StrategySweepRequest extends StrategyBacktestRequest {
  /**
   * 参数网格列表，每个元素可覆盖 strategy_params 或 signal_params。
   * 后端对每个组合独立跑一次回测，返回 `rows: SweepRow[]`。
   */
  grid: Array<{
    strategy_params?: Record<string, unknown>
    signal_params?: Record<string, unknown>
  }>
}

/**
 * 启动 Walk-Forward 验证的请求体（`POST /api/strategy/walkforward/run`）。
 *
 * 以滚动窗口在时间轴上依次训练→测试，评估策略参数的时序稳定性。
 */
export interface StrategyWalkForwardRequest extends StrategyBacktestRequest {
  /** 每折训练窗口长度（自然日数），建议 ≥ 90 */
  train_days: number
  /** 每折测试窗口长度（自然日数），建议 ≥ 30 */
  test_days: number
}

/**
 * 参数扫描 / Walk-Forward 单行结果（`result.rows` 列表元素）。
 *
 * 用于在扫描结果表格中对比各参数组合的回测表现。
 */
export interface SweepRow {
  /** 当前行覆盖的参数字典 */
  params: Record<string, unknown>
  /** 总收益率（小数，如 0.25 = 25%） */
  total_return: number
  /** 夏普比率 */
  sharpe_ratio: number
  /** 最大回撤百分比（小数，如 0.15 = 15%） */
  max_ddpercent: number
  /** 总净收益（元） */
  total_net_pnl: number
  /** 总成交笔数 */
  trade_count: number
}

// Re-export backtest result types from alpha (reuse existing types)
export type { BacktestStatistics, BacktestResultPayload } from './alpha'

File: backend/aitrade/models/strategy.py
"""
规则策略回测/扫参/WalkForward 请求模型（Phase 2）。

风格对齐 models/alpha.py：pydantic v2，字段注释中文，类型用 X | None。
"""

from __future__ import annotations

from datetime import date
from typing import Any

from pydantic import BaseModel, Field, model_validator


class StrategyCost(BaseModel):
    """回测成本与制度约束（A 股规则策略默认值）。

    commission_rate / stamp_duty / slippage 均为小数比例（如 0.0003 = 万三）；
    T+1 默认开启，复现 A 股当日买入不可当日卖出的交易规则。
    """

    commission_rate: float = Field(default=0.0003, description="单边佣金率")
    stamp_duty: float = Field(default=0.0005, description="卖出印花税率")
    slippage: float = Field(default=0.0005, description="每笔不利滑点率")
    t_plus1: bool = Field(default=True, description="T+1 卖出限制（规则策略默认开启，贴近 A 股现实）")


class StrategyBacktestRequest(BaseModel):
    """单次规则策略回测请求体。

    ``signal_source`` 对应注册表名（如 "etf_momentum"），由 ``api/strategy.py`` 路由
    解析后分派给相应信号源工厂；``strategy_params`` 透传给策略（如 rebalancing_topk 的 top_k）。
    日期合法性由 ``_check_date_order`` 校验（start 须严格早于 end）。
    """

    signal_source: str = Field(description="注册表名，如 'etf_momentum'")
    signal_params: dict[str, Any] = Field(default_factory=dict, description="透传给信号源工厂的参数")
    strategy_name: str = Field(default="rebalancing_topk", description="已注册的策略名")
    strategy_params: dict[str, Any] = Field(default_factory=dict, description="策略参数（top_k/n_drop/min_days 等）")
    interval: str = Field(default="d", description="K 线周期")
    start: date = Field(description="回测起始日（含）")
    end: date = Field(description="回测截止日（含）")
    capital: int = Field(default=1_000_000, description="初始资金")
    cost: StrategyCost = Field(default_factory=StrategyCost, description="成本与制度约束")

    @model_validator(mode="after")
    def _check_date_order(self) -> StrategyBacktestRequest:
        """校验 start 严格早于 end；否则抛 ValueError（Pydantic 会将其转为 422）。"""
        if self.start >= self.end:
            raise ValueError(f"start ({self.start}) 必须早于 end ({self.end})")
        return self


class StrategySweepRequest(StrategyBacktestRequest):
    """网格扫参请求：grid 中每项覆盖 strategy_params 和/或 signal_params。

    继承 ``StrategyBacktestRequest`` 全部字段；``grid`` 中每个 dict 作为 override
    与基础参数合并后运行独立回测，结果按顺序返回。grid 上限 50 项（防服务端任务泛洪）。
    """

    grid: list[dict[str, Any]] = Field(
        description="参数覆盖列表，每项含 strategy_params 和/或 signal_params 的 override，上限 50"
    )

    @model_validator(mode="after")
    def _check_grid_size(self) -> StrategySweepRequest:
        """校验 grid 不超过 50 项（防止服务端任务泛洪）。"""
        if len(self.grid) > 50:
            raise ValueError(f"grid 不能超过 50 项，当前 {len(self.grid)} 项")
        return self


class StrategyWalkForwardRequest(StrategyBacktestRequest):
    """Walk-Forward 验证请求（规则策略无训练，train 窗作参数观察期）。

    规则策略不涉及参数拟合，``train_days`` 作为参数稳定性观察期使用；
    ``test_days`` 为每折 OOS 测试窗口长度，折数由总时间范围自动推算。
    """

    train_days: int = Field(default=180, description="训练窗（参数观察期）天数")
    test_days: int = Field(default=60, description="测试窗天数；逐 test 窗跑回测")
